dropshot/src/pagination.rs

```diff
@@ -0,0 +1,569 @@
+ * happen if a malicious client finds some way to cause requests to hang or take
+ * a very long time.
+ *
+ * [1]: https://cloud.google.com/apis/design/design_patterns#list_pagination
+ * [2]: https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/
+ */
+
+use crate as dropshot; // XXX needed for ExtractedParameter below
+use crate::error::HttpError;
+use base64::URL_SAFE;
+use dropshot_endpoint::ExtractedParameter;
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use serde::Serialize;
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::num::NonZeroU64;
+
+/**
+ * Client's view of a page of results from a paginated API
+ */
+#[derive(Debug, Deserialize, JsonSchema, Serialize)]
+pub struct ClientPage<ItemType> {
+    /** token to be used in pagination params for the next page of results */
+    pub next_page: Option<String>,
+    /** list of items on this page of results */
+    pub items: Vec<ItemType>,
+}
+
+/**
+ * Pagination interface for simple cases where a collection is sorted in one
+ * direction by one Serializable field (called the marker)
+ *
+ * This is simpler to use than `PaginationParams` directly, but it's less
+ * flexible.  See examples/pagination-marker.rs for details.
+ */
+pub type MarkerPaginator<FieldType> =
+    PaginationParams<MarkerScanMode, MarkerPageSelector<FieldType>>;
+
+/**
+ * Internal type used as the `ScanMode` type for `MarkerPaginator`.
+ */
+#[derive(Debug, Deserialize, ExtractedParameter)]
+#[serde(rename_all = "kebab-case")]
+pub enum MarkerScanMode {
+    ByMarkerAsc,
+}
+
+/**
+ * Internal type used as the `PageSelector` type for `MarkerPaginator`.
+ */
+#[derive(Debug, Deserialize, ExtractedParameter, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum MarkerPageSelector<F> {
+    Marker(F),
+}
+
+/*
+ * Generic pagination infrastructure
+ * XXX need to better understand the trait bounds / serde bounds here.
+ */
+
+/**
+ * Query parameters provided by clients when scanning a paginated collection
+ *
+ * See `PaginatedResource` for more information.
+ */
+#[derive(Debug, Deserialize, ExtractedParameter)]
+#[serde(bound(
+    deserialize = "PageSelector: DeserializeOwned, ScanMode: DeserializeOwned"
+))]
+pub struct PaginationParams<ScanMode, PageSelector>
+where
+    ScanMode: Debug + Send + Sync + 'static,
+    PageSelector: Serialize + Send + Sync + 'static,
+{
+    /**
+     * Specifies either how the client wants to begin a new scan or how to
+     * resume a previous scan.  This field is flattened by serde, so see the
+     * variants of [`WhichPage`] to see which fields actually show up in the
+     * serialized form.
+     */
+    #[serde(flatten)]
+    pub page_params: WhichPage<ScanMode, PageSelector>,
+
+    /**
+     * Optional client-requested limit on page size.  Consumers should use
+     * [`dropshot::RequestContext::page_limit()`] to access this value.
+     */
+    pub(crate) limit: Option<NonZeroU64>,
```

Comment:
should the limit have a limit? We can express this in OpenAPI. I do wonder if there's a single limit that makes sense for all paginated items (100? 200? 1000?) or if we'd want to let different items have different limits (seems like just more rope to hang ourselves with).

Reply:
That'd be neat.  Any idea how to represent that in the source?

Reply:
We can do it with a custom `impl JsonSchema` like I did for `ApiName` here: https://github.com/oxidecomputer/oxide-api-prototype/blob/master/src/api_model.rs#L183

This is for a string rather than an integer, but I think you get the point.

I'd be happy to contribute that if it would be helpful.

Reply:
For the record: we discussed this in person and decided it's not really part of the API contract, but a runtime parameter of the server.